main.go

```diff
@@ -1,173 +1,253 @@
+	UsePyron           bool   `env:"USE_PYRON,required"`
+	ResourceURL        string `env:"RESOURCE_URL"`
+	XSSLCertHash       string `env:"X_SSL_CERT_HASH"`
+	InjectCertMode     bool   `env:"INJECT_CERT_MODE,required"`
+}
+
+func (c Config) NewClientConfig() acp.Config {
 	var (
-		serverPort int
-		acpClient  acp.Client
-		err        error
+		redirectURL *url.URL
+		err         error
 	)
 
-	flag.Parse()
-	if serverPort, err = strconv.Atoi(*port); err != nil {
-		log.Fatalln(err)
+	if redirectURL, err = url.Parse(fmt.Sprintf("http://%v:%v/callback", c.RedirectHost, c.PORT)); err != nil {
+		log.Fatalf("failed to get callback url from host %v", err)
+	}
+
+	return acp.Config{
+		ClientID:     c.ClientID,
+		RedirectURL:  redirectURL,
+		TokenURL:     c.TokenEndpoint,
+		AuthorizeURL: c.AuthorizeEndpoint,
+		IssuerURL:    c.IssuerURL,
+		CertFile:     c.CertPath,
+		KeyFile:      c.KeyPath,
+		RootCA:       c.RootCA,
+		Scopes:       []string{"openid"},
 	}
+}
 
-	acpOAuthConfig = acp.Config{
-		RedirectURL: fmt.Sprintf("http://%v:%v/callback", *redirectHost, serverPort),
-		ClientID:    *clientID,
-		Scopes:      []string{"openid"},
-		AuthURL:     fmt.Sprintf("%v/oauth2/authorize", *issuerURL),
-		TokenURL:    fmt.Sprintf("%v/oauth2/token", *issuerURL),
-		PKCEEnabled: *pkceEnabled,
+func LoadConfig() (config Config, err error) {
+	if err = env.Parse(&config); err != nil {
+		return config, err
 	}
 
-	if acpClient, err = acp.NewClient(*serverCertPath, *certPath, *keyPath, acpOAuthConfig); err != nil {
-		log.Fatalln(err)
+	if config.WellKnownURL, err = url.Parse(config.WellKnown); err != nil {
+		log.Fatalf("failed to parse wellknown as url %v", err)
 	}
 
-	handler := http.NewServeMux()
-	handler.HandleFunc("/callback", callback(acpClient))
-	handler.HandleFunc("/login", login)
-
-	server := &http.Server{
-		Addr:    fmt.Sprintf("%v:%v", *host, serverPort),
-		Handler: handler,
-		TLSConfig: &tls.Config{
-			MinVersion:               tls.VersionTLS12,
-			CurvePreferences:         []tls.CurveID{tls.CurveP521, tls.CurveP384, tls.CurveP256},
-			PreferServerCipherSuites: true,
-			CipherSuites: []uint16{
-				tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
-				tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
-				tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
-				tls.TLS_RSA_WITH_AES_256_CBC_SHA,
+	return config, err
+}
+
+func loadTemplates() (*template.Template, error) {
+	return template.ParseFiles(layoutFiles()...)
+}
+
+func layoutFiles() []string {
+	files, err := filepath.Glob("templates/*.html")
+	if err != nil {
+		log.Fatal(err)
+	}
+	return files
+}
+
+type Server struct {
+	Config       Config
+	Client       acp.Client
+	HttpClient   *http.Client
+	SecureCookie *securecookie.SecureCookie
+}
+
+func NewServer() (Server, error) {
+	var (
+		client = Server{}
+		err    error
+	)
+
+	if client.Config, err = LoadConfig(); err != nil {
+		return client, errors.Wrapf(err, "failed to load config")
+	}
+
+	client.Config.fetchEndpointURLs()
```

Comment:
I'd avoid function that mutates, instead I'd do sth like:

```
func NewServer() (Server, error) {
	var (
		server       = Server{}
		endpoints    WellKnownEndpoints
		clientConfig acp.Config
		err          error
	)

	if server.Config, err = LoadConfig(); err != nil {
		return server, errors.Wrapf(err, "failed to load config")
	}

	if server.HttpClient, err = newHTTPClient(server.Client, server.Config); err != nil {
		return server, errors.Wrapf(err, "failed to get http client")
	}

	if endpoints, err = getServerEndpoints(server.HttpClient, server.Config); err != nil {
		return server, errors.Wrapf(err, "failed to get server endpoints")
	}

	if clientConfig, err = buildClientConfig(server.Config, endpoints); err != nil {
		return server, errors.Wrapf(err, "failed to build client config")
	}

	if server.Client, err = acp.New(clientConfig); err != nil {
		return server, errors.Wrapf(err, "failed to init acp client")
	}
...
}
```

Please note that if you already have http client, you can use it to call well-known endpoint instead of default http client.